lucene/core/src/java/org/apache/lucene/index/SegmentCommitInfo.java

```diff
@@ -388,4 +399,17 @@ public SegmentCommitInfo clone() {
   final int getDelCount(boolean includeSoftDeletes) {
     return includeSoftDeletes ? getDelCount() + getSoftDelCount() : getDelCount();
   }
+
+  private void generationAdvanced() {
+    sizeInBytes = -1;
+    id = null;
+  }
+
+  public byte[] getId() {
+    if (id == null) {
+      // we advanced a generation - need to generate a new ID
+      id = StringHelper.randomId();
```

Comment:
Do we need some thread safety here?

Reply:
yeah good question, as far as I can tell we never read or write any of the member vars on this class unless we hold a lock that protects it. But it's a mess so I guess we should. Yet, if we do that we should make every method on this class synced no?

Reply:
I agree it's messy.  It's just that this sounds like a simple getter w/o side effect, yet, it does have an important and surprising side effect of setting the `id`.  It would be bad if two threads called it at once and got two different `id`s back.

If we really believe all callers are already sync'd (by IW's monitor lock), can we 1) add a comment stating this assumption, and 2) fix the assignment to do something like this instead:

```
  byte[] newID = StringHelper.randomId();
  assert id == null;
  id = newID;
```

?

That is still best effort, but might catch us in tests if we have cases where two threads really are invoking `getId` at once.

Or, maybe we can change this to assign the `id` at the same time as the SCI was changed (e.g. `delGen` increased), instead of at `get` time?

Reply:
I changed this to roll over to a new ID each time the generation is incremented. I don't really like that we call `StringHelper.randomId()` over and over again. I wonder if it would be better to set a new ID from the outside just before we clone in order to commit instead of doing it inside SegmentCommitInfo. This would make things more explicit IMO